src/crypto/CHIPCryptoPAL.h

```diff
@@ -1285,5 +1285,28 @@ enum class MatterOid
  **/
 CHIP_ERROR ExtractDNAttributeFromX509Cert(MatterOid matterOid, const ByteSpan & certificate, uint16_t & id);
 
+class AesCtrKeyContext
+{
+public:
+    virtual ~AesCtrKeyContext()                                                      = default;
+    virtual CHIP_ERROR Encrypt(MutableByteSpan & plaintext, const ByteSpan & nonce)  = 0;
+    virtual CHIP_ERROR Decrypt(MutableByteSpan & ciphertext, const ByteSpan & nonce) = 0;
+};
+
+class AesCcmKeyContext
+{
+public:
+    virtual ~AesCcmKeyContext()                                  = default;
+    virtual CHIP_ERROR Encrypt(MutableByteSpan & plaintext, const ByteSpan & aad, const ByteSpan & nonce,
+                               MutableByteSpan & out_mic)        = 0;
+    virtual CHIP_ERROR Decrypt(MutableByteSpan & ciphertext, const ByteSpan & aad, const ByteSpan & nonce,
+                               const ByteSpan & mic)             = 0;
+    virtual CHIP_ERROR DerivePrivacyKey(AesCtrKeyContext *& out) = 0;
```

Comment:
Reference to pointer is always a bit confusing. Perhaps we need to replace `DerivePricacyKey` with PrivacyEncrypt/PrivacyDecrypt methods that use the derived key internally. The impl can then derive the key once if not already derived. This avoids having to manage memory for the AesCtrKeyContext (which still can be used internally by an impl)

Reply:
I agree. I have created a single interface with methods to decrypt/encrypt the message, and decrypt/encrypt the privacy fields. The privacy methods will internally generate the privacy key as needed.